package/csac_daemon/src/csac_daemon.c

```diff
@@ -0,0 +1,307 @@
+/*
+ * Copyright (C) 2018 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#include <getopt.h>
+#include <string.h>
+#include <unistd.h>
+#include <termios.h>
+
+#include <libpiksi/logging.h>
+#include <libpiksi/sbp_pubsub.h>
+#include <libpiksi/loop.h>
+#include <libpiksi/settings.h>
+
+#include <libsbp/sbp.h>
+#include <libsbp/piksi.h>
+#include <libsbp/system.h>
+
+#include "serial_utils.h"
+
+#define PROGRAM_NAME "csac_daemon"
+
+#define SBP_SUB_ENDPOINT "ipc:///var/run/sockets/internal.pub" /* SBP Internal Out */
+#define SBP_PUB_ENDPOINT "ipc:///var/run/sockets/internal.sub" /* SBP Internal In */
+
+#define SBP_FRAMING_MAX_PAYLOAD_SIZE (255u)
+#define CSAC_TELEM_UPDATE_INTERVAL (1000u)
+#define HEADER_SEND_DIVISOR (10)
+#define POLL_SLEEP_MICROSECONDS (100000)
+
+u8 *port_name = NULL;
+u8 *command_string = NULL;
+bool csac_telemetry_enabled = false;
+
+struct csac_ctx_s {
+  sbp_pubsub_ctx_t *sbp_ctx;
+  serial_port_t *port;
+  char csac_telem[SBP_FRAMING_MAX_PAYLOAD_SIZE];
+  char csac_header[SBP_FRAMING_MAX_PAYLOAD_SIZE];
+};
+
+static bool csac_daemon_enabled()
+{
+  return csac_telemetry_enabled;
+}
+
+static void usage(char *command)
+{
+  printf("Usage: %s\n", command);
+
+  puts("\nMain options");
+  puts("\t--serial-port <serial port for csac daemon>");
+}
+
+static int parse_options(int argc, char *argv[])
+{
+  enum { OPT_ID_SERIAL_PORT = 1, OPT_ID_AT_COMMAND = 2 };
+
+  const struct option long_opts[] = {
+    {"serial-port", required_argument, 0, OPT_ID_SERIAL_PORT},
+    {0, 0, 0, 0},
+  };
+
+  int opt;
+  while ((opt = getopt_long(argc, argv, "", long_opts, NULL)) != -1) {
+    switch (opt) {
+    case OPT_ID_SERIAL_PORT: {
+      port_name = (u8 *)optarg;
+    } break;
+
+    default: {
+      puts("Invalid option");
+      return -1;
+    } break;
+    }
+  }
+
+  if (port_name == NULL) {
+    puts("Missing port");
+    return -1;
+  }
+
+  return 0;
+}
+
+static int read_until_cr_lf(int fd, char *outbuf, size_t maxlen)
+{
+  int nbytes;   /* Number of bytes read */
+  char *bufptr; /* Current char in buffer */
+  bufptr = outbuf;
+  char *eob = outbuf + maxlen;
+  while ((nbytes = read(fd, bufptr, (size_t)(eob - bufptr) - 1)) > 0) {
+    bufptr += nbytes;
+    if ((size_t)(bufptr - outbuf) > 2) {
+      if (bufptr[-1] == '\n' || bufptr[-1] == '\r') {
```

Comment:
I think we could use `fgetc` to make this code simpler if it's possible for us to use `fdopen` on the device (using `setbuf` and `setvbuf` you can configure the buffering for the stream).  Using `[-1]` to "index" strings in C is pretty non-standard/odd.

Reply:
Also, in master we have the `file_read_string` helper function in libpiksi: https://github.com/swift-nav/piksi_buildroot/blob/master/package/libpiksi/libpiksi/src/util.c#L94 -- but this would require that the read be terminated with `\n` (the current function terminates if either `\r` OR `\n` is encoutered).

Reply:
planning to do when I port this to master next week.